src/realm/object-store/collection.cpp

```diff
@@ -0,0 +1,157 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include <realm/object-store/collection.hpp>
+#include <realm/object-store/shared_realm.hpp>
+#include <realm/object-store/object_schema.hpp>
+#include <realm/object-store/object_store.hpp>
+#include <realm/object-store/results.hpp>
+
+namespace realm::object_store {
+
+Collection::OutOfBoundsIndexException::OutOfBoundsIndexException(size_t r, size_t c)
+    : std::out_of_range(util::format("Requested index %1 greater than max %2", r, c - 1))
+    , requested(r)
+    , valid_count(c)
+{
+}
+
+Collection::Collection() noexcept {}
+
+Collection::Collection(std::shared_ptr<Realm> r, const Obj& parent_obj, ColKey col)
+    : m_realm(std::move(r))
+    , m_type(ObjectSchema::from_core_type(col) & ~PropertyType::Collection)
+    , m_coll_base(parent_obj.get_collection_ptr(col))
+{
+}
+
+Collection::Collection(std::shared_ptr<Realm> r, const CollectionBase& coll)
+    : m_realm(std::move(r))
+    , m_type(ObjectSchema::from_core_type(coll.get_col_key()) & ~PropertyType::Collection)
+    , m_coll_base(coll.clone())
+{
+}
+
+Collection::~Collection() {}
+
+Collection::Collection(const Collection&) = default;
+Collection& Collection::operator=(const Collection&) = default;
+Collection::Collection(Collection&&) = default;
+Collection& Collection::operator=(Collection&&) = default;
+
+bool Collection::is_valid() const
+{
+    if (!m_realm)
+        return false;
+    m_realm->verify_thread();
+    if (!m_realm->is_in_read_transaction())
+        return false;
+    return m_coll_base->is_attached();
+}
+
+ObjKey Collection::get_parent_object_key() const
+{
+    verify_attached();
+    return m_coll_base->get_key();
+}
+
+ColKey Collection::get_parent_column_key() const
+{
+    verify_attached();
+    return m_coll_base->get_col_key();
+}
+
+TableKey Collection::get_parent_table_key() const
+{
+    verify_attached();
+    return m_coll_base->get_table()->get_key();
+}
+
+static StringData object_name(Table const& table)
+{
+    return ObjectStore::object_type_for_table_name(table.get_name());
+}
+
+
+void Collection::validate(const Obj& obj) const
+{
+    if (!obj.is_valid())
+        throw std::invalid_argument("Object has been deleted or invalidated");
+    auto target = m_coll_base->get_target_table();
```

Comment:
I guess we would probably like to extend this for TypedLink columns and Mixed?

Reply:
Added a FIXME